internal/controller/clusterorder_resources.go

```diff
@@ -0,0 +1,142 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	cloudkitv1alpha1 "github.com/innabox/cloudkit-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *ClusterOrderReconciler) newNamespace(ctx context.Context, instance *cloudkitv1alpha1.ClusterOrder) (*appResource, error) {
+	namespaceName := instance.GetClusterReferenceNamespace()
+	if namespaceName == "" {
+		namespaceName = generateNamespaceName(instance)
+		if namespaceName == "" {
+			return nil, fmt.Errorf("failed to generate namespace name")
+		}
+	}
+
+	namespace := &corev1.Namespace{
+		TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "Namespace"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   namespaceName,
+			Labels: commonLabelsFromOrder(instance),
+		},
+	}
+
+	instance.SetClusterReferenceNamespace(namespaceName)
+
+	mutateFn := func() error {
+		ensureCommonLabels(instance, namespace)
+		return nil
+	}
+
+	return &appResource{
+		namespace,
+		mutateFn,
+		true,
+	}, nil
+}
+
+func (r *ClusterOrderReconciler) newServiceAccount(ctx context.Context, instance *cloudkitv1alpha1.ClusterOrder) (*appResource, error) {
+	namespaceName := instance.GetClusterReferenceNamespace()
+	if namespaceName == "" {
+		return nil, fmt.Errorf("unable to retrieve required information from spec.clusterReference")
+	}
+
+	serviceAccountName := defaultServiceAccountName
+	instance.SetClusterReferenceServiceAccountName(serviceAccountName)
+
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceAccountName,
+			Namespace: namespaceName,
+			Labels:    commonLabelsFromOrder(instance),
+		},
+	}
+
+	mutateFn := func() error {
+		ensureCommonLabels(instance, sa)
+		return nil
+	}
+
+	return &appResource{
+		sa,
+		mutateFn,
+		true,
+	}, nil
+}
+
+func (r *ClusterOrderReconciler) newAdminRoleBinding(ctx context.Context, instance *cloudkitv1alpha1.ClusterOrder) (*appResource, error) {
+	namespaceName := instance.GetClusterReferenceNamespace()
+	serviceAccountName := instance.GetClusterReferenceServiceAccountName()
+	if namespaceName == "" || serviceAccountName == "" {
+		return nil, fmt.Errorf("unable to retrieve required information from spec.clusterReference")
+	}
+
+	roleBindingName := defaultRoleBindingName
+
+	subjects := []rbacv1.Subject{
+		{
+			Kind:      "ServiceAccount",
+			Name:      serviceAccountName,
+			Namespace: namespaceName,
+		},
+	}
+
+	roleref := rbacv1.RoleRef{
+		APIGroup: rbacv1.GroupName,
+		Kind:     "ClusterRole",
+		Name:     "admin",
+	}
+
+	roleBinding := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleBindingName,
+			Namespace: namespaceName,
+			Labels:    commonLabelsFromOrder(instance),
+		},
+		Subjects: subjects,
```

Comment:
Typically anything outside the ObjectMeta should be added in the mutateFn, which always gets called regardless of the operation being a create or update. CreateOrUpdate will try to Get the object using its client, which if successful will overwrite contents of the object passed in, but IIRC optional or nil values won't get unset. So if you set desired state here, you can potentially end up with an object in the mutateFn that is a mix between the desired state and the actual current state. We want to ensure that it always represents just the current state when it pre-exists.

You are doing that now in the mutateFn, so you could just ditch these two lines since they're redundant anyway.